src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,545 @@
+
+                    // e.g. `e is { p: 1 } x && x is C` => `is C { p: 1 } x`
+                    (RecursivePatternSyntax { Type: null } recursive, TypePatternSyntax type)
+                        => recursive.WithType(type.Type),
+
+                    // e.g. `e is { p: 1 } x && x is C { p: 2 }` => `e is C { p: 1, p: 2 } x`
+                    (RecursivePatternSyntax left, RecursivePatternSyntax { Designation: null } right)
+                        when (left.Type is null || right.Type is null) &&
+                             (left.PositionalPatternClause is null || right.PositionalPatternClause is null)
+                        => left
+                            .WithType(left.Type ?? right.Type)
+                            .WithPositionalPatternClause(left.PositionalPatternClause ?? right.PositionalPatternClause)
+                            .WithPropertyPatternClause(Concat(left.PropertyPatternClause, right.PropertyPatternClause)),
+
+                    // In any other case, we fallback to an `and` pattern.
+                    // UNDONE: This may result in a few unused variables which should be removed in a later pass.
+                    _ => BinaryPattern(AndPattern, containingPattern.Parenthesize(), generatedPattern.Parenthesize()),
+                };
+            }
+
+            static PatternSyntax AddSubpattern(PatternSyntax containingPattern, SubpatternSyntax subpattern)
+            {
+                return containingPattern switch
+                {
+                    // e.g. `case var x when x.p is 1` => `case { p: 1 } x`
+                    VarPatternSyntax p => RecursivePattern(type: null, subpattern, p.Designation),
+
+                    // e.g. `case Type x when x.p is 1` => `case Type { p: 1 } x`
+                    DeclarationPatternSyntax p => RecursivePattern(p.Type, subpattern, p.Designation),
+
+                    // e.g. `case { p: 1 } x when x.q is 2` => `case { p: 1, q: 2 } x`
+                    RecursivePatternSyntax p => p.AddPropertyPatternClauseSubpatterns(subpattern),
+
+                    // We've already checked that the designation is contained in any of the above pattern forms.
+                    var p => throw ExceptionUtilities.UnexpectedValue(p)
+                };
+            }
+
+            static PropertyPatternClauseSyntax? Concat(PropertyPatternClauseSyntax? left, PropertyPatternClauseSyntax? right)
+            {
+                if (left is null || right is null)
+                    return left ?? right;
+                var leftSubpatterns = left.Subpatterns.GetWithSeparators();
+                if (leftSubpatterns.Any() && !leftSubpatterns.Last().IsToken)
+                    leftSubpatterns = leftSubpatterns.Add(Token(CommaToken));
+                var rightSubpatterns = right.Subpatterns.GetWithSeparators();
+                var list = new SyntaxNodeOrTokenList(leftSubpatterns.Concat(rightSubpatterns));
```

Comment:
i looked, but couldn't find a helper for thsi anywhere... sorry!

Reply:
It should be an extension on SyntaxNodeOrTokenList but I couldn't find any existing extension class for it (where it should go?)